apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Transaction.java

```diff
@@ -221,9 +239,63 @@ public void incrementReferences() {
 
     public void decrementReferences() {
         final int referenceCount = this.references.decrementAndGet();
-        super.decrementReferences();
+        if (logger.isTraceEnabled()) {
+            logger.trace("decrement references to {} ({})", this, referenceCount);
+        }
         if (referenceCount == 0) {
             tracer.recycle(this);
         }
     }
+
+    void incrementTimer(@Nullable String type, @Nullable String subtype, long duration) {
+        if (!collectBreakdownMetrics || type == null || finished) {
+            return;
+        }
+        final Labels.Mutable spanType = labelsThreadLocal.get();
+        spanType.resetState();
+        Timer timer = spanTimings.get(spanType.spanType(type).spanSubType(subtype));
+        if (timer == null) {
+            timer = new Timer();
+            Timer racyTimer = spanTimings.putIfAbsent(spanType.immutableCopy(), timer);
+            if (racyTimer != null) {
+                timer = racyTimer;
+            }
+        }
+        timer.update(duration);
+        if (finished) {
+            // in case end()->trackMetrics() has been called concurrently
+            // don't leak timers
```

Comment:
Still not thread safe:
```
Thread A:
      AbstractSpan#end() -> Transaction#doEnd -> trackMetrics() // all timers are reset, finished == false

Thread B:
      AbstractSpan#end() -> Span#doEnd // creates an put a Timer, doesn't clean because finished == false

Thread A:
      AbstractSpan#end() -> Transaction#doEnd (continue) 
      AbstractSpan#end() -> sets finished = true 
```

Reply:
will set `finished = true` earlier

Reply:
should be fixed now